pkg/csi/service/wcpguest/controller.go

```diff
@@ -575,13 +575,28 @@ func controllerPublishForBlockVolume(ctx context.Context, req *csi.ControllerPub
 			if err == nil || time.Now().After(timeout) {
 				break
 			}
+			if err != nil {
+				log.Errorf("failed to update virtualmachine. Err: %v", err)
+			}
 			virtualMachine = &vmoperatortypes.VirtualMachine{}
 			if err := c.vmOperatorClient.Get(ctx, vmKey, virtualMachine); err != nil {
 				msg := fmt.Sprintf("failed to get VirtualMachines for the node: %q. Error: %+v", req.NodeId, err)
 				log.Error(msg)
 				return nil, csifault.CSIInternalFault, status.Errorf(codes.Internal, msg)
 			}
 			log.Debugf("Found virtualMachine instance for node: %q", req.NodeId)
+			for _, volume := range virtualMachine.Status.Volumes {
```

Comment:
I don't quite get why this checking the Status, when this loop is adding the volume to the Spec. This loop is duplicated from like 549-556 so I think this reads & flows a lot easier if this is done once after the Spec is updated - if it needs to be.  Something akin to:

```
if !isVolumePresentInSpec {
  timeout := ...
  for {
      // Loop until we add volume to Spec, or it is already present
  }
  if err != nil {
    // we timedout
   return ...
}

isVolumeAttached := false
for _, volume := range virtualMachine.Status.Volumes {
...
}

if !isVolumeAttached {
  // watch & wait until Status is updated
}
```

Reply:
@bryanv
I am checking the status right after getting the new`virtualMachine` object, before patching it in the loop.

```
for {
			virtualMachine.Spec.Volumes = append(virtualMachine.Spec.Volumes, vmvolumes)
			err := c.vmOperatorClient.Update(ctx, virtualMachine)
                        virtualMachine = &vmoperatortypes.VirtualMachine{}
			if err := c.vmOperatorClient.Get(ctx, vmKey, virtualMachine); err != nil {
                        ..
                         if the volume is attached by some other prior call to this VM then exit with loop and return
}
```

this eliminates continuous updates in the loop when the volume is already attached by different call.
let me see if I can make the logic simple.

Reply:
@bryanv I have simplified the logic as you suggested?
Can you take a look again?

Reply:
@bryanv lock is also removed with this commit - https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/1746/commits/2d4be5eedde0dd34dd8702225d834617a53a911d 

Thanks for the review.